prov/util/src/util_attr.c

```diff
@@ -71,41 +71,83 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
 	}
 }
 
-char *ofi_strdup_less_prefix(char *name, char *prefix)
+/*
+char *ofi_strdup_head(const char *str)
 {
-	return strdup(name + strlen(prefix) + 1);
+	char *delim;
+	delim = strchr(str, OFI_NAME_DELIM);
+	return delim ? strndup(str, delim - str) : strdup(str);
 }
 
-char *ofi_strdup_add_prefix(char *name, char *prefix)
+char *ofi_strdup_tail(const char *str)
 {
-	char *prefix_name;
-	char *base = "";
-	ssize_t size;
-	int ret;
-
-	if (name)
-		base = name;
-
-	size = snprintf(NULL, 0, "%s_%s", prefix, base) + 1;
-	if (size < 0)
-		return NULL;
+	char *delim;
+	delim = strchr(str, OFI_NAME_DELIM);
+	return delim ? strup(delim + 1) : strdup(str);
+}
+*/
 
-	prefix_name = calloc(size, sizeof(*prefix_name));
-	if (!prefix_name)
-		return NULL;
+char *ofi_strdup_append(const char *head, const char *tail)
+{
+	char *str;
+	size_t len;
+
+	len = strlen(head) + strlen(tail) + 2;
+	str = malloc(len);
+	if (str)
+		sprintf(str, "%s%c%s", head, OFI_NAME_DELIM, tail);
+	return str;
+}
 
-	ret = snprintf(prefix_name, size, "%s_%s", prefix, base);
-	if (ret < 0 || ret > size)
-		goto err;
+static int ofi_has_util_prefix(const char *str)
+{
+	return !strncasecmp(str, OFI_UTIL_PREFIX, strlen(OFI_UTIL_PREFIX));
+}
 
-	return prefix_name;
+const char *ofi_util_name(const char *str, size_t *len)
+{
+	char *delim;
+
+	delim = strchr(str, OFI_NAME_DELIM);
+	if (delim) {
+		if (ofi_has_util_prefix(delim + 1)) {
+			*len = strlen(delim + 1);
+			return delim + 1;
+		} else if (ofi_has_util_prefix(str)) {
+			*len = delim - str;
+			return str;
+		}
+	} else if (ofi_has_util_prefix(str)) {
+		*len = strlen(str);
+		return str;
+	}
+	*len = 0;
+	return NULL;
+}
 
-err:
-	free(prefix_name);
+const char *ofi_core_name(const char *str, size_t *len)
+{
+	char *delim;
+
+	delim = strchr(str, OFI_NAME_DELIM);
+	if (delim) {
+		if (!ofi_has_util_prefix(delim + 1)) {
```

Comment:
As implemented by libfabric - yes.  An app could specify these in either order, though I don't document that this is supported.